rules_bazel/java/java_diff_test.bzl

```diff
@@ -1,87 +1,84 @@
     )
 
     ctx.actions.run_shell(
         inputs = inputs,
-        outputs = [output],
-        arguments = [test_class_name],
+        outputs = [goldens_output_zip],
+        arguments = [test_class],
         tools = [test_runner],
-        command = command,
+        command = generate_goldens_script,
     )
 
-junit_output_zip = rule(
-    attrs = {
-        "test_class_name": attr.string(mandatory=True),
-        "srcs": attr.label_list(
-            allow_files = True,
-            mandatory = True,
-        ),
-        "test_runner": attr.label(
-            mandatory = True,
-            executable = True,
-            cfg = "host",
-        ),
-    },
-    outputs = {
-        "output": "%{name}.zip",
-    },
-    implementation = _junit_output_impl,  
-)
-    
-def _overwrite_golden_impl(ctx):
-    script_content = """
+    # Overwrite the goldens.
+    golden_update_script_content = """
     #!/bin/bash
     cd ${{BUILD_WORKSPACE_DIRECTORY}}
-    unzip -ao {unit_test_results} -d src/test/java
+    unzip -ao {goldens_output_zip} -d src/test/java
     """.format(
-        unit_test_results = ctx.file.unit_test_results.path,
+        goldens_output_zip = goldens_output_zip.path,
     )
     ctx.actions.write(
-        output = ctx.outputs.bin,
-        content = script_content,
+        output = ctx.outputs.golden_update_script,
+        content = golden_update_script_content,
         is_executable = True,
     )
-    return [DefaultInfo(executable = ctx.outputs.bin)]
-
+    return [DefaultInfo(executable = ctx.outputs.golden_update_script)]
 
 overwrite_golden = rule(
     attrs = {
-        "unit_test_results": attr.label(
+        "test_class": attr.string(mandatory = True),
+        "srcs": attr.label_list(
+            allow_files = True,
             mandatory = True,
-            allow_single_file = True),
+        ),
+        "test_runner": attr.label(
+            mandatory = True,
+            executable = True,
+            cfg = "host",
+        ),
     },
     outputs = {
-        "bin": "%{name}.sh",
+        "goldens_output_zip": "%{name}.zip",
+        "golden_update_script": "%{name}.sh",
     },
     executable = True,
     implementation = _overwrite_golden_impl,
 )
 
-def golden_update(name, test_class_name, srcs):
-    junit_output_name = "%s_output" % name
-    junit_output_zip(
-        name = junit_output_name,
-        test_class_name = test_class_name,
-        test_runner = "//:golden_update_junit_runner",
+def golden_update(
+        name,
+        srcs,
+        test_class,
+        data = [],
+        deps = []):
+    golden_junit_runner_name = "%s_junit_runner" % name
+    native.java_binary(
+        name = golden_junit_runner_name,
         srcs = srcs,
+        data = data,
+        main_class = "com.google.api.generator.test.framework.SingleJUnitTestRunner",
+        deps = ["//src/test/java/com/google/api/generator/test/framework:junit_runner"] + deps,
     )
```

Comment:
Can you explain a little bit why the `def junit_output_zip ` is replaced by the `native.java_binary` and it will make the build running before running overwrite_golden?  Thanks